ostree/ostree_dest.go

```
@@ -0,0 +1,245 @@
+		fmt.Sprintf("--add-metadata-string=docker.size=%d", blob.Size),
+		fmt.Sprintf("--branch=%s", ostreeBranch),
+		fmt.Sprintf("--tree=dir=%s", destinationPath)).Run()
+}
+
+func (d *ostreeImageDestination) HasBlob(info types.BlobInfo) (bool, int64, error) {
+	branch := fmt.Sprintf("ociimage/%s", strings.Replace(info.Digest.Hex(), "sha256:", "", 1))
+	output, err := exec.Command("ostree", "show", "--print-metadata-key=docker.size", branch).CombinedOutput()
+	if err != nil {
+		if  bytes.Index(output, []byte("not found")) >= 0 || bytes.Index(output, []byte("No such")) >= 0 {
+			return false, -1, nil
+		} else {
+			return false, -1, err
+		}
+	}
+	size, err := strconv.ParseInt(strings.Trim(string(output), "'\n"), 10, 64)
+	if err != nil {
+		return false, -1, err
+	}
+
+	return true, size, nil
+}
+
+func (d *ostreeImageDestination) ReapplyBlob(info types.BlobInfo) (types.BlobInfo, error) {
+	return info, nil
+}
+
+func (d *ostreeImageDestination) PutManifest(manifest []byte) error {
+	d.manifest = strings.Replace(string(manifest), "\n", " ", -1)
+
+	if err := json.Unmarshal(manifest, &d.schema); err != nil {
+		return err
+	}
+
+	manifestPath := d.ref.manifestPath()
+	if err := ensureParentDirectoryExists(manifestPath); err != nil {
+		return err
+	}
+
+	return ioutil.WriteFile(manifestPath, manifest, 0644)
+}
+
+func (d *ostreeImageDestination) PutSignatures(signatures [][]byte) error {
+	if err := ensureParentDirectoryExists(d.ref.signaturePath(0)); err != nil {
+		return err
+	}
+
+	for i, sig := range signatures {
+		signaturePath := d.ref.signaturePath(i)
+		if err := ioutil.WriteFile(signaturePath, sig, 0644); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *ostreeImageDestination) Commit() error {
+	defer func() {
+		os.RemoveAll(encodeOStreeRef(d.ref.image))
+	}()
+
+	for _, layer := range d.schema.LayersDescriptors {
+		hash := strings.Replace(layer.Digest.Hex(), "sha256:", "", 1)
+		blob := d.blobs[hash]
+		if blob == nil {
+			continue
```

Silently ignoring failures seems unsafe